src/app/components/fruit-table/fruit-table.component.ts

```diff
@@ -8,11 +8,106 @@ import {FruitTableViewModel} from './fruit-table-view-model';
   providers: [FruitTableViewModel]
 })
 export class FruitTableComponent implements OnInit {
-  columnsToDisplay = ['id', 'name', 'genus', 'calories'];
+  columnsToDisplay = ['id', 'name', 'genus', 'calories', 'carbohydrates', 'sugar'];
 
   constructor(public viewModel: FruitTableViewModel) {
 
   }
+  dataSource = this.viewModel.fruitData$
+  filteredDataSource = this.dataSource.value||this.dataSource
+
+  fruitFilter(event: Event) {
+    console.log('event', event)
+    const filterBy = (event.target as HTMLInputElement).value.toLowerCase();
+    const dataTable = this.dataSource.value
+    let resultTable = []
+    console.log('filter term:',filterBy)
+    console.log('local dataTable',dataTable)
+    console.log('filteredData',this.filteredDataSource)
+    dataTable.forEach((fruitInfo) => {
+      // console.log ('for each:',fruitInfo)
+      if  (fruitInfo.genus.toLowerCase().includes(filterBy)
+          ||fruitInfo.name.toLowerCase().includes(filterBy)
+          ||fruitInfo.family.toLowerCase().includes(filterBy)
+          ||fruitInfo.order.toLowerCase().includes(filterBy)) {
+            resultTable.push(fruitInfo)
+      }
+      this.filteredDataSource=resultTable
+    })
+    console.log('filtered results',resultTable)
+  }
+
+  fruitSort (event: Event) {
+    const sortBy = (event.target as HTMLInputElement).value
+    console.log('sort fruits by:', sortBy)
+    let dataTable = this.dataSource
```

Comment:
I think part of the reason that the filter and sort don't work nicely together is because this line is using the original datasource and not the filtered one. If you switch to the filteredDataSource here that might fix the issue.